### DISCUSSION TITLE
How to get dataset through the git LFS pointer? Example?

### DISCUSSION THREAD

**Discussion by BirdyLiu6471023** [git-lfs/git-lfs#5112] [Q&A] [answered]
There is a pointer link of git LFS, and I search in the website how to access, no one could give me obvious explanation. When I get a "oid sha256:......", what should I do? Is there any step-by-step instructions? Do not just give me a code like use "git lfs fetch" or 'git lfs pull". Please be specific, where, what and how? Thank! AND I HAVE ALREADY DOWNLOAD THE git LFS. **DO NOT** tell me how to install it!

**Comment by chrisd8088** [comment]
Well, I'm not quite sure how to help you, but I might suggest reading through this comprehensive introductory [tutorial](https://www.atlassian.com/git/tutorials/git-lfs).

If your question is "how do I manually download a file's contents given a Git LFS pointer file" then I think the key point to understand is that you're unlikely (depending on your Git storage provider) to be able to do that on the command line without a lot of effort.  That's because the Git LFS client communicates with the Git LFS service over a batch API, which is [documented](https://github.com/git-lfs/git-lfs/blob/52d47f085ba5ae534dd26ba76d010a537ddbdc8e/docs/api/batch.md) in detail if you're interested, and as part of that exchange it requests download URLs for a given set of objects.  The URLs provided by the server may include various authorization features without which you would not be able to download the objects directly.

But I'm not sure if I'm answering the question; you might want to include some command line examples to show what you're trying to do.

**Reply by BirdyLiu6471023** [reply]
Thanks for your reply! The error shows below: 
`Changes to be committed:
  (use "git rm --cached <file>..." to unstage)
	new file:   .gitattributes
	new file:   Archive.zip
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.dbf
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.prj
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.qpj
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.shp
	new file:   Archive/sepher2.0_16_18_final_nyc_metro_nta_disp_fin.shx

(base) MacbookdeMBP-3:new repo macbook$ git lfs track "*.csv"
Tracking "*.csv"
(base) MacbookdeMBP-3:new repo macbook$ git lfs pull
Could not pull

Errors logged to '/Users/macbook/Desktop/new repo/.git/lfs/logs/20220913T140509.173711.log'.
Use `git lfs logs last` to view the log.
(base) MacbookdeMBP-3:new repo macbook$ git lfs fetch
Could not fetch

Errors logged to '/Users/macbook/Desktop/new repo/.git/lfs/logs/20220913T140530.632829.log'.
Use `git lfs logs last` to view the log.`

**Comment by chrisd8088** [comment, marked as answer]
That looks like Git LFS can't use Git to resolve the `HEAD` reference.  What do you get from `git rev-parse HEAD --symbolic-full-name HEAD`?  Is it possible your `.git/HEAD` file points to a branch name that doesn't exist?

**Reply by BirdyLiu6471023** [reply]
Thanks for your help! That is very helpful and useful! Because I am a totally new in git but need to get the data from this, which is very tough for me. By the way, I want to know if I want to process the data I fetched out, what should I do? After fetching, it shows like this: 
` (base) dyn-129-236-217-10:dataset_test macbook$ git lfs fetch`
`fetch: Fetching reference refs/heads/birdy`
`(base) dyn-129-236-217-10:dataset_test macbook$`

**Reply by chrisd8088** [reply]
That seems like it succeeded ... if so, you should find that the LFS pointer files in your working tree have been replaced with their full contents.

**Reply by BirdyLiu6471023** [reply]
If you mean "replace", opening the original file, it should come up with the full contents. But when I open the file, it seems that it still the pointer. Is there any other steps I need to do? Thanks!

**Reply by chrisd8088** [reply]
You could try `git lfs checkout`; that should populate any pointer files in your working directory with their contents -- assuming the client was able to fetch those contents.

I was a little too quick to reply before: `git lfs fetch` will fetch the LFS objects but will store them only in the private `.git/lfs/objects` directory, whereas `git lfs pull` will fetch them *and* populate your working tree with their contents.  So `git lfs pull` is like `git lfs fetch` followed by `git lfs checkout`.

**Reply by BirdyLiu6471023** [reply]
Thanks for your reply! I really appreciate it! However, I want to know do you mean if I open my original .csv file through the excel, it would automatically become into full content or something else?  And I want to clarify that I only have .csv file with three lines like, without github link: 
`version http:.... `
`oid sha256:.....`
`size 786 ...` 

And I want to attach all of my code in bash here: 
`
(base) MacbookdeMBP-3:data_shape macbook$ git status
On branch main
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv

nothing added to commit but untracked files present (use "git add" to track)
(base) MacbookdeMBP-3:data_shape macbook$ git add .
(base) MacbookdeMBP-3:data_shape macbook$ git status
On branch main
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv

(base) MacbookdeMBP-3:data_shape macbook$ git lfs track "*.csv"
Tracking "*.csv"
(base) MacbookdeMBP-3:data_shape macbook$ git lfs pull
`

After this pull, I checked my original .csv file, it still contains those three line above. Thanks for your help!

**Reply by chrisd8088** [reply]
I confess it's a little hard for me to suggest solutions because it's unclear to me exactly what steps you have followed to get into this state, or what you are trying to do.

One specific point is that you will want to do `git lfs track "*.csv"` not `git lfs track ".csv"`.  The former will let Git LFS match  all files with that extension, while the latter will only match a single file named exactly `.csv`, and so won't match your `sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv` file.

Another specific point is that would you normally want to do a `git commit` after the `git add` and `git lfs track` steps, not a `git lfs pull`, because the point is that you are adding a file to your Git repository and committing it along with a Git LFS rule to match its filename and handle it as a Git LFS object.

Again, I would strongly suggest working through some normal Git tutorials as well as the excellent Git LFS [tutorial](https://www.atlassian.com/git/tutorials/git-lfs).

For instance, I would expect you to have *added* the `.csv` file to your repository in the first place like this:
```
$ git lfs track '*.csv'
$ git add sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv
$ git commit -m 'creating my CSV file'
$ git push
```

When you run `git commit` in that sequence it will, assuming Git LFS is installed, run a filter process which turns the file into a Git LFS object in the repository.  And then `git push` will (1) push your changes to the remote repository and (2) push the Git LFS file to the remote's Git LFS storage.  You would not want to `git pull` or `git lfs pull` at this point because that's pulling data *from* the remote while you're trying to *update* the remote.

But, since you have a file which sounds like it is a Git LFS pointer file -- the one with the three lines in it -- that must have come from somewhere; either you created it locally or fetched it from your remote repository.

Here is what would be helpful:
1) The output of a `git lfs env` command.
2) The output of a `git lfs track` command.
3) The output of a `git lfs ls-files` command.
4) The output of a `git show HEAD:sepher2.0_16_18_final_nyc_metro_nta_disp_fin.csv` command (or whatever the filename is of the file with three lines).
